src/test/java/org/jenkinsci/plugins/gitclient/GitAPITestCase.java

```diff
@@ -2666,10 +2666,37 @@ public void test_submodule_update_with_error() throws Exception {
             w.git.submoduleUpdate().execute();
             fail("Did not throw expected submodule update exception");
         } catch (GitException e) {
-            assertThat(e.getMessage(), containsString("Command \"git submodule update modules/ntp\" returned status code 1"));
+            assertThat(e.getMessage(), containsString("Command \"git submodule update --checkout --force modules/ntp\" returned status code 1"));
         }
     }
 
+    @NotImplementedInJGit
+    public void test_submodule_update_with_dirty_index() throws Exception {
+        w.git.clone_().url(localMirror()).execute();
+        w.git.checkout().ref("origin/tests/getSubmodules").execute();
+        // fetch the submodules a first time to initialize the directories.
+        w.git.submoduleUpdate().execute();
+        // create a new file and commit the submodule
+        File ntpRepo = w.file("modules/ntp");
+        WorkingArea submoduleWorkingArea = new WorkingArea(ntpRepo).init();
+        submoduleWorkingArea.touch("test_file");
+        submoduleWorkingArea.git.add("test_file");
+        submoduleWorkingArea.git.commit("submod-commit1");
+        // add the changed submodule to the repo
+        w.git.add("modules/ntp");
+        w.git.commit("commit1");
+        String head = w.git.revParse("HEAD").name();
+        // revert to the previous commit and clean the submodule
+        w.launchCommand("git", "reset", "HEAD~");
+        w.git.submoduleUpdate().execute();
+        // manually create the test_file again, which does not exist in this revision of the submodule
+        submoduleWorkingArea.touch("test_file");
+        // revert back to the previous HEAD
+        w.launchCommand("git", "reset", head);
+        // Now try and launch the submodule update again. Without --force, this will fail.
+        w.git.submoduleUpdate().execute();
```

Comment:
Please assert the content of the repository in some way at this point as well.  It is great that you've added a submodule test.  It will be even better if there are more assertions included in the test.